components/arm/trossen/trossen.go

```diff
@@ -239,6 +254,83 @@ func (a *Arm) JointPositions(ctx context.Context, extra map[string]interface{})
 	return &pb.JointPositions{Values: positions}, nil
 }
 
+// DoCommand handles incoming gripper requests.
+func (a *Arm) DoCommand(ctx context.Context, cmd map[string]interface{}) (map[string]interface{}, error) {
+	name, ok := cmd["command"]
+	if !ok {
+		return nil, errors.New("missing 'command' value")
+	}
+	switch name {
+	case TrossenGripperClose:
+		grabbed, err := a.grab(ctx)
+		return map[string]interface{}{"grabbed": grabbed}, err
+	case TrossenGripperOpen:
+		return nil, a.open(ctx)
+	default:
+		return nil, errors.Errorf("no such command: %s", name)
+	}
+}
+
+// Opens the gripper. Accessible via DoCommand and called by the
+// Trossen gripper implementation (through DoCommand).
+func (a *Arm) open(ctx context.Context) error {
+	ctx, done := a.opMgr.New(ctx)
+	defer done()
+	a.moveLock.Lock()
+	defer a.moveLock.Unlock()
+	err := a.Joints["Gripper"][0].SetGoalPWM(150)
+	if err != nil {
+		return err
+	}
+
+	// We don't want to over-open
+	atPos := false
+	for !atPos {
+		var pos int
+		pos, err = a.Joints["Gripper"][0].PresentPosition()
+		if err != nil {
+			return err
+		}
+		if pos < 2800 {
+			if !utils.SelectContextOrWait(ctx, 50*time.Millisecond) {
+				return ctx.Err()
+			}
+		} else {
+			atPos = true
+		}
+	}
+	err = a.Joints["Gripper"][0].SetGoalPWM(0)
+	return err
+}
+
+// Closes the gripper. Accessible via DoCommand and called by the
+// Trossen gripper implementation (through DoCommand).
+func (a *Arm) grab(ctx context.Context) (bool, error) {
+	_, done := a.opMgr.New(ctx)
+	defer done()
+	a.moveLock.Lock()
+	defer a.moveLock.Unlock()
+	err := a.Joints["Gripper"][0].SetGoalPWM(-350)
+	if err != nil {
+		return false, err
+	}
+	err = servo.WaitForMovementVar(a.Joints["Gripper"][0])
+	if err != nil {
+		return false, err
+	}
+	pos, err := a.Joints["Gripper"][0].PresentPosition()
+	if err != nil {
+		return false, err
+	}
+	didGrab := true
+
+	// If servo position is less than 1500, it's closed and we grabbed nothing
+	if pos < 1500 {
+		didGrab = false
+	}
+	return didGrab, nil
+}
+
 // Stop is unimplemented for trossen.
 func (a *Arm) Stop(ctx context.Context, extra map[string]interface{}) error {
```

Comment:
[request] add `a.opMgr.CancelRunning(ctx, nil)`

Reply:
or the corresponding clean version of that code.

Reply:
should this be added to all of the commands? (as in should none of them be blocking?)

Reply:
it's command dependent and we need to figure out that functionality. I just know it necessarily has to be implemented in `Stop`. This is sufficient for now and I'm resolving this comment. Thank you!